aptible/resource_app.go

```diff
@@ -52,47 +54,217 @@ func resourceApp() *schema.Resource {
+			},
+			"post_scale_up_cooldown_seconds": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     60,
+				Description: "Cooldown period in seconds after a scale-up event.",
+			},
+			"post_scale_down_cooldown_seconds": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     300,
+				Description: "Cooldown period in seconds after a scale-down event.",
+			},
+			"post_release_cooldown_seconds": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     300,
+				Description: "Cooldown period in seconds after a release event.",
+			},
+			"mem_cpu_ratio_r_threshold": {
+				Type:     schema.TypeFloat,
+				Default:  4.0,
+				Optional: true,
+			},
+			"mem_cpu_ratio_c_threshold": {
+				Type:     schema.TypeFloat,
+				Default:  2.0,
+				Optional: true,
+			},
+			"mem_scale_up_threshold": {
+				Type:        schema.TypeFloat,
+				Optional:    true,
+				Default:     0.9,
+				Description: "The memory usage threshold for scaling up.",
+			},
+			"mem_scale_down_threshold": {
+				Type:        schema.TypeFloat,
+				Optional:    true,
+				Default:     0.75,
+				Description: "The memory usage threshold for scaling down.",
+			},
+			"minimum_memory": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     2048,
+				Description: "The minimum memory allocation in MB.",
+			},
+			"maximum_memory": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "The maximum memory allocation in MB.",
+			},
+			"min_cpu_threshold": {
+				Type:        schema.TypeFloat,
+				Optional:    true,
+				Description: "The minimum CPU utilization threshold for scaling.",
+			},
+			"max_cpu_threshold": {
+				Type:        schema.TypeFloat,
+				Optional:    true,
+				Description: "The maximum CPU utilization threshold for scaling.",
+			},
+			"min_containers": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "The minimum number of containers for scaling.",
+			},
+			"max_containers": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "The maximum number of containers for scaling.",
+			},
+			"scale_up_step": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     1,
+				Description: "The number of containers to add in each scale-up event.",
+			},
+			"scale_down_step": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     1,
+				Description: "The number of containers to remove in each scale-down event.",
 			},
 		},
 	}
 }
 
+func validateServiceSizingPolicy(ctx context.Context, d *schema.ResourceDiff, _ interface{}) error {
+	services := d.Get("service").(*schema.Set).List()
+	for _, service := range services {
+		serviceMap := service.(map[string]interface{})
+		if policy, ok := serviceMap["service_sizing_policy"]; ok {
+			policies := policy.(*schema.Set).List() // should only be one, but yes it's a Set
```

Comment:
I know these type casts are everywhere in this codebase but they are pretty dangerous.

https://go.dev/tour/methods/16

Not a blocker just calling it out if we do run into issues.

Reply:
Thanks, I actually use that trick elsewhere!

Over there though policy is guaranteed to be a schema.Set due to the terraform schema